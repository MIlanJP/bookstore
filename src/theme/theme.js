import { createMuiTheme } from '@material-ui/core/styles';
import purple from '@material-ui/core/colors/purple';
import yellow from '@material-ui/core/colors/yellow';
import red from '@material-ui/core/colors/red';
const fontForBookName="'Lato'"
const theme = createMuiTheme({
  palette: {
    primary: {
        main:red[500]
    },
    secondary: {
      main: yellow[500],
    },
    themeColor:{
        
    }
  },
  typography: {
    bookName: {
      fontFamily: fontForBookName,
      fontWeight: 700,
    //   fontSize: 10,
      textTransform: "none",
    },
    authorName: {
        fontFamily: fontForBookName,
        // fontWeight: 700,
        fontSize: 13.5,
        textTransform: "none",
      },
      priceColumn:{
        fontFamily: fontForBookName,
        fontWeight: 700,
        // fontSize: 13.5,
        position: "relative",
        top:'-10px',
        textTransform: "none",  
      },
      buttonText:{
        fontFamily: fontForBookName,
        fontWeight: 300,
        fontSize: 13,
        textTransform: "none", 
      },    
        wishListButton:{
        fontFamily: fontForBookName,
        fontWeight: 700,
        fontSize: 13.5,
        textTransform: "none", 
      },
      Latofont:{
        fontFamily: fontForBookName,
        fontWeight: 700,
      } ,     
      placeOrder:{
        fontFamily: fontForBookName,
        fontWeight: 200,
      }

},
});

export default theme;