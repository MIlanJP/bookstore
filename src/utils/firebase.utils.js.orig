import firebase from 'firebase';
import firebase1 from 'firebase/app';

import admin from 'firebase-admin'

const firebaseConfig = {
    apiKey: "AIzaSyCF8I46ftSKLpqRBZ1xPIWZXbkA4I2zK04",
    authDomain: "bookstore-5636e.firebaseapp.com",
    databaseURL: "https://bookstore-5636e.firebaseio.com",
    projectId: "bookstore-5636e",
    storageBucket: "bookstore-5636e.appspot.com",
    messagingSenderId: "540723071433",
    appId: "1:540723071433:web:f2065f00acaf7ed9e24018",
    measurementId: "G-NWV2SXZW3D"
  };

  firebase.initializeApp(firebaseConfig)

  export const auth=firebase.auth();
  export const fireStore=firebase.firestore();
  const provider=new firebase.auth.GoogleAuthProvider();
  provider.setCustomParameters({prompt:'select_account'})
  // export const createUserProfileDocument=async(userAuth,additionalData)=>{
  //   // if(userAuth==='') return 
  //   const userRef=fireStore.doc('users/PPuI3rXmthxyGDqGOmML')
  //   const snapshot=await userRef.get()
  //   console.log(JSON.s snapshot)
  //    fireStore.collection('users').get().then( 
  //     (data)=>{
  //       // console.log(data.docs)
   
  //     }
  //   )
  // }
  export const signInWithGoogle=()=>{auth.signInWithPopup(provider)}
export default firebase;

