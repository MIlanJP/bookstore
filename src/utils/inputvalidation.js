
class Validation {
    vallidateName(name) {
      return /^[A-Z]+[a-z1-9]*$/.test(name) && name.length > 4;
    }
    
  }
  
  export default new Validation();