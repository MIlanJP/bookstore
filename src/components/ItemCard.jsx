import { Button, Card, CardMedia } from "@material-ui/core";
import { makeStyles } from "@material-ui/styles";
import {addBookToCart} from '../redux'
import _ from "lodash";
import {useDispatch,useSelector} from 'react-redux'
import {updateCartQuantity} from '../redux'
import {useHistory} from 'react-router'
import React, { useState, useEffect } from "react";
import {fireStore} from '../utils/firebase.utils'
function ItemCard(props) {
const [displayAddedButton,setDisplayAddedButton]=useState(_.some( props.userData.itemsList, ["id",props.book.id]))
const history = useHistory();
const items = useSelector((state) => state.cardquantity.items);

  const useStyle = makeStyles((theme) => ({
    mainLayout: {
      // background:'black',
      border: "black",
      height: "315px",
      width: "220px",
      marginLeft:'10px',
    },
    bookImage: {
      height: "171px",
      width: "220px",
      paddingBottom: "2px",
      background: "rgb(245,245,245)",
    },
    imageOfBook: {
      marginTop: "7%",
      height: "80%",
    },
    bookName: {
      ...theme.typography.bookName,
      paddingLeft: "25px",
      whiteSpace: "nowrap",
      textOverflow: "ellipsis",
      textAlign: "left",
      overflow: "hidden",
      paddingRight: "20px",
    },
    authorName: {
      ...theme.typography.authorName,
      color: "#9D9D9D",
      paddingLeft: "25px",
      textAlign: "left",
      lineHeight: ".8rem",
      position: "relative",
      top: "-14px",
      display: "flex",
    },
    priceColumn: {
      ...theme.typography.priceColumn,
      paddingLeft: "25px",
      textAlign: "left",
      top: "-21px",
      position: "relative",
    },
    addToCartButton: {
      ...theme.typography.buttonText,
      background: "rgb(160,48,55)",
      color: "white",
      //   marginRight:"10px",
      padding: "4px 10px 5px 10px",
      "&:hover": {
        ...theme.typography.buttonText,
        background: "rgb(160,48,55)",
        color: "white",
      },
    },
    buttonsColumn: {
      position: "relative",
      top: "-20px",
      display: "flex",
      flexDirection: "row",
      justifyContent: "space-evenly",
    },
    wishListButton: {
      ...theme.typography.wishListButton,
      boxShadow: " 0px 0px  3px 0px black",
      padding: "3px 10px 3px 10px",
      width: props.book.quantity > 0 ? "" : "90%",

      //   marginLeft:'10px'
    },
    outOfStock: {
      position: "absolute",
      background: "rgba(255, 255, 255, 0.95)",
      boxShadow: "0px 0px 5px 0px black",
      height: "40px",
      width: "184px",
      display: "flex",
      justifyContent: "center",
      alignItems: "center",
      marginTop: "70px",
      fontWeight: "600",
      marginLeft: "17px",
    },
    addedTOCartButton: {
      ...theme.typography.wishListButton,
      boxShadow: " 0px 0px  3px 0px black",
      padding: "3px 10px 3px 10px",
      width: "90%",
      color:'white',
      position: "relative",
      top: "-20px",
      background: "rgb(51,113,181)",
      "&:hover": {
        background: "rgb(51,113,181)",
      },
    },
  }));
  const dispatch= useDispatch()
  const classes = useStyle();


  return (
    <Card className={classes.mainLayout}>
      <div className={classes.bookImage}>
        {props.book.quantity > 0 ? null : (
          <div  data-testid="out-of-stock-tag"    className={classes.outOfStock}>OUT OF STOCK</div>
        )}
        <img
          className={classes.imageOfBook}
          src={props.book.image}
          alt={props.book.title}
        />
      </div>

      <p className={classes.bookName} 
      data-testid="bookTitle"
      >
        {props.book.title}
      </p>
      <span className={classes.authorName} data-testid="authorName"  > by {props.book.author}</span>
      <p className={classes.priceColumn} data-testid="priceColumn"  >Rs {props.book.price}</p>
      {displayAddedButton ? (
        <Button className={classes.addedTOCartButton}
        data-testid="added-to-cart-button" 
        >ADDED TO CART</Button>
      ) : (
        <div className={classes.buttonsColumn}>
          {props.book.quantity > 0 ? (
            <Button
            data-testid="add-to-bag-button" 
              className={classes.addToCartButton}
              onClick={(e) => {
               

               let usersData=props.userData
               let lists=props.userData.itemsList
              //  let booksInCart=parseInt(props.userData.booksInCart)+1;
               let book=props.book;
               book.quantity=1;
               lists.push(book)
               usersData.itemsList=[...lists]
               usersData.booksInCart=parseInt(items)+1
               const userRef = fireStore.doc(`users/${props.userData.id}`);
               userRef.get().then(async (snapShot) => {
                 try {
                   await userRef.update({
                     booksInCart: parseInt(items)+1,
                     itemsList: lists,
                   });
                   props.setUserData(usersData);
                   dispatch(addBookToCart(props.book))
               setDisplayAddedButton(true)
               dispatch(updateCartQuantity(parseInt(items)+1))

                 } catch (err) {}
          
               });
              }}
            >
              ADD TO BAG
            </Button>
          ) : null}

          <Button  data-testid="wishlistButton"    className={classes.wishListButton}>WISHLIST</Button>
        </div>
      )}
    </Card>
  );
}

export default ItemCard;
