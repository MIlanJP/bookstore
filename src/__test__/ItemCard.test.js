import React from "react";
import { BrowserRouter } from "react-router-dom";
import { ThemeProvider } from "@material-ui/styles";
import testingLibrary from "@testing-library/react";

import theme from "../theme/theme";

import { render, fireEvent, screen, cleanup } from "../utils/testUtils";
import ItemCard from "../components/ItemCard";

const book=  {
    id: "11",
    author: "Dan Brown",
    title: "Angels And Demons",
    image:
      "http://books.google.com/books/content?id=d5xgYw4Ts0gC&printsec=frontcover&img=1&zoom=5",
    quantity: "15",
    price: "612",
    description:
      "*INCLUDES A SNEAK PREVIEW OF ORIGINTHE NEW THRILLER BY DAN BROWN: OUT 3RD OCTOBER. PRE-ORDER TODAY* --------------------------------------------------------------------------------------------------- CERN InstituteSwitzerland: a world-renowned scientist is found brutally murdered with a mysterious symbol seared onto his chest. The VaticanRome: the College of Cardinals assembles to elect a new pope. Somewhere beneath theman unstoppable bomb of terrifying power relentlessly counts down to oblivion. In a breathtaking race against timeHarvard professor Robert Langdon must decipher a labyrinthine trail of ancient symbols if he is to defeat those responsible - the Illuminatia secret brotherhood presumed extinct for nearly four hundred yearsreborn to continue their deadly vendetta against their most hated enemythe Catholic Church",
  }

  const outOfStockBook=  {
    id: "11",
    author: "Dan Brown",
    title: "Angels And Demons",
    image:
      "http://books.google.com/books/content?id=d5xgYw4Ts0gC&printsec=frontcover&img=1&zoom=5",
    quantity: "0",
    price: "612",
    description:
      "*INCLUDES A SNEAK PREVIEW OF ORIGINTHE NEW THRILLER BY DAN BROWN: OUT 3RD OCTOBER. PRE-ORDER TODAY* --------------------------------------------------------------------------------------------------- CERN InstituteSwitzerland: a world-renowned scientist is found brutally murdered with a mysterious symbol seared onto his chest. The VaticanRome: the College of Cardinals assembles to elect a new pope. Somewhere beneath theman unstoppable bomb of terrifying power relentlessly counts down to oblivion. In a breathtaking race against timeHarvard professor Robert Langdon must decipher a labyrinthine trail of ancient symbols if he is to defeat those responsible - the Illuminatia secret brotherhood presumed extinct for nearly four hundred yearsreborn to continue their deadly vendetta against their most hated enemythe Catholic Church",
  }

const bookList = [
  {
    id: "1",
    author: "Chetan Bhagat",
    title: "The Girl in Room 105",
    image:
      "http://books.google.com/books/content?id=GHt_uwEACAAJ&printsec=frontcover&img=1&zoom=5",
    quantity: "12",
    price: "193",
    description:
      "Hi I'm Keshavand my life is screwed. I hate my job and my girlfriend left me. Ahthe beautiful Zara. Zara is from Kashmir. She is a Muslim. And did I tell you my family is a bitwelltraditional? Anywayleave that. Zara and I broke up four years ago. She moved on in life. I didn''t. I drank every night to forget her. I calledmessagedand stalked her on social media. She just ignored me. Howeverthat nighton the eve of her birthdayZara messaged me. She called me overlike old timesto her hostel room 105. I shouldn''t have gonebut I did... and my life changed forever. This is not a love story. It is an unlove story.From the author of Five Point Someone and 2 Statescomes a fast-pacedfunny and unputdownable thriller about obsessive love and finding purpose in life against the backdrop of contemporary India",
  },
  {
    id: "2",
    author: "Rujuta Divekar",
    title: "Indian Superfoods",
    image:
      "http://books.google.com/books/content?id=4oFoDwAAQBAJ&printsec=frontcover&img=1&zoom=5",
    quantity: "13",
    price: "495",
    description:
      "Forget about acacia seeds and goji berries. The secret foods for healthvitality and weight loss lie in our own kitchens and backyards. Top nutritionist Rujuta Diwekar talks you through the ten Indian superfoods that will completely transform you'",
  },
  {
    id: "3",
    author: "Dan Brown",
    title: "Angels And Demons",
    image:
      "http://books.google.com/books/content?id=d5xgYw4Ts0gC&printsec=frontcover&img=1&zoom=5",
    quantity: "14",
    price: "218",
    description:
      "*INCLUDES A SNEAK PREVIEW OF ORIGINTHE NEW THRILLER BY DAN BROWN: OUT 3RD OCTOBER. PRE-ORDER TODAY* --------------------------------------------------------------------------------------------------- CERN InstituteSwitzerland: a world-renowned scientist is found brutally murdered with a mysterious symbol seared onto his chest. The VaticanRome: the College of Cardinals assembles to elect a new pope. Somewhere beneath theman unstoppable bomb of terrifying power relentlessly counts down to oblivion. In a breathtaking race against timeHarvard professor Robert Langdon must decipher a labyrinthine trail of ancient symbols if he is to defeat those responsible - the Illuminatia secret brotherhood presumed extinct for nearly four hundred yearsreborn to continue their deadly vendetta against their most hated enemythe Catholic Church",
  },
];

const userData = {
  name: "",
  email: "",
  booksInCart: 0,
  itemsList: [...bookList],
};

describe("Test Books Container", () => {
  afterEach(cleanup);
  it("Renders the connected app with initialState without error", () => {
    const {getByTestId}=render(
      <ThemeProvider theme={theme}>
        <BrowserRouter>
          <ItemCard userData={userData} book={bookList[0]} />
        </BrowserRouter>
      </ThemeProvider>,
      { initialState: { user: "Redux User", userData, cardquantity: 0 } }
      
    );
    expect(getByTestId("authorName" )).toHaveTextContent("Chetan Bhagat")
    expect(getByTestId("priceColumn" )).toHaveTextContent("Rs 193")
    expect(getByTestId("bookTitle" )).toHaveTextContent("The Girl in Room 105")
    
    expect(screen.getByText("ADDED TO CART")).toBeInTheDocument();
  });


  it("Renders the connected app if book present in cart", () => {
   const {asFragment}= render(
      <ThemeProvider theme={theme}>
        <BrowserRouter>
          <ItemCard userData={userData} book={bookList[0]} />
        </BrowserRouter>
      </ThemeProvider>,
      { initialState: { user: "Redux User", userData, cardquantity: 0 } }
    );
    expect(asFragment()).toMatchSnapshot()
  });

  
  it("Renders the connected app with if book not present in cart", () => {
    const {asFragment}= render(
       <ThemeProvider theme={theme}>
         <BrowserRouter>
           <ItemCard userData={userData} book={book} />
         </BrowserRouter>
       </ThemeProvider>,
       { initialState: { user: "Redux User", userData, cardquantity: 0 } }
     );
     expect(asFragment()).toMatchSnapshot()
   });

   it("Renders Add To cart button if book is not present in cart", () => {
    const {getByTestId}= render(
       <ThemeProvider theme={theme}>
         <BrowserRouter>
           <ItemCard userData={userData} book={book} />
         </BrowserRouter>
       </ThemeProvider>,
       { initialState: { user: "Redux User", userData, cardquantity: 0 } }
     );
     expect(getByTestId("add-to-bag-button" )).toHaveTextContent("ADD TO BAG")
   });

   
   it("Should not render the Add to cart button when book is in cart", () => {
    const {queryAllByTestId}= render(
       <ThemeProvider theme={theme}>
         <BrowserRouter>
           <ItemCard userData={userData} book={bookList[0]} />
         </BrowserRouter>
       </ThemeProvider>,
       { initialState: { user: "Redux User", userData, cardquantity: 0 } }
     );
     expect(queryAllByTestId("add-to-bag-button" )).toHaveLength(0)
   });


   it("Renders Add To cart button if book is not present in cart", () => {
    const {getByTestId}= render(
       <ThemeProvider theme={theme}>
         <BrowserRouter>
           <ItemCard userData={userData} book={book} />
         </BrowserRouter>
       </ThemeProvider>,
       { initialState: { user: "Redux User", userData, cardquantity: 0 } }
     );
     expect(getByTestId("wishlistButton" )).toHaveTextContent("WISHLIST")
   });

   it("Should not render the wishlist button when book is in cart", () => {
    const {queryAllByTestId}= render(
       <ThemeProvider theme={theme}>
         <BrowserRouter>
           <ItemCard userData={userData} book={bookList[0]} />
         </BrowserRouter>
       </ThemeProvider>,
       { initialState: { user: "Redux User", userData, cardquantity: 0 } }
     );
     expect(queryAllByTestId("wishlistButton" )).toHaveLength(0)
   });


  it("Renders Added to Cart button if book is present in the cart", () => {
    const {getByTestId}= render(
       <ThemeProvider theme={theme}>
         <BrowserRouter>
           <ItemCard userData={userData} book={bookList[0]} />
         </BrowserRouter>
       </ThemeProvider>,
       { initialState: { user: "Redux User", userData, cardquantity: 0 } }
     );
     expect(getByTestId("added-to-cart-button" )).toHaveTextContent("ADDED TO CART")
    });

      it("Does not render the ADDED TO CART button when book not in cart", () => {
    const {queryAllByTestId}= render(
       <ThemeProvider theme={theme}>
         <BrowserRouter>
           <ItemCard userData={userData} book={book} />
         </BrowserRouter>
       </ThemeProvider>,
       { initialState: { user: "Redux User", userData, cardquantity: 0 } }
     );
     expect(queryAllByTestId("added-to-cart-button" )).toHaveLength(0)
   });



   it("Renders out of stock tag when quantity of book is zero", () => {
    const {getByTestId}= render(
       <ThemeProvider theme={theme}>
         <BrowserRouter>
           <ItemCard userData={userData} book={outOfStockBook} />
         </BrowserRouter>
       </ThemeProvider>,
       { initialState: { user: "Redux User", userData, cardquantity: 0 } }
     );
     expect(getByTestId("out-of-stock-tag" )).toHaveTextContent("OUT OF STOCK")
   });

   it("Does not render out of stock when book is in stock", () => {
    const {queryAllByTestId}= render(
       <ThemeProvider theme={theme}>
         <BrowserRouter>
           <ItemCard userData={userData} book={book} />
         </BrowserRouter>
       </ThemeProvider>,
       { initialState: { user: "Redux User", userData, cardquantity: 0 } }
     );
     expect(queryAllByTestId("out-of-stock-tag" )).toHaveLength(0)
   });



});
