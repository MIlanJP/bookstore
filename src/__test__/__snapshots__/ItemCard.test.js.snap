// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Books Container Renders the connected app if book present in cart 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-mainLayout-12 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="makeStyles-bookImage-13"
    >
      <img
        alt="The Girl in Room 105"
        class="makeStyles-imageOfBook-14"
        src="http://books.google.com/books/content?id=GHt_uwEACAAJ&printsec=frontcover&img=1&zoom=5"
      />
    </div>
    <p
      class="makeStyles-bookName-15"
      data-testid="bookTitle"
    >
      The Girl in Room 105
    </p>
    <span
      class="makeStyles-authorName-16"
      data-testid="authorName"
    >
       by Chetan Bhagat
    </span>
    <p
      class="makeStyles-priceColumn-17"
      data-testid="priceColumn"
    >
      Rs 193
    </p>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-addedTOCartButton-22"
      data-testid="added-to-cart-button"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        ADDED TO CART
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`Test Books Container Renders the connected app with if book not present in cart 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root makeStyles-mainLayout-23 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="makeStyles-bookImage-24"
    >
      <img
        alt="Angels And Demons"
        class="makeStyles-imageOfBook-25"
        src="http://books.google.com/books/content?id=d5xgYw4Ts0gC&printsec=frontcover&img=1&zoom=5"
      />
    </div>
    <p
      class="makeStyles-bookName-26"
      data-testid="bookTitle"
    >
      Angels And Demons
    </p>
    <span
      class="makeStyles-authorName-27"
      data-testid="authorName"
    >
       by Dan Brown
    </span>
    <p
      class="makeStyles-priceColumn-28"
      data-testid="priceColumn"
    >
      Rs 612
    </p>
    <div
      class="makeStyles-buttonsColumn-30"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-addToCartButton-29"
        data-testid="add-to-bag-button"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          ADD TO BAG
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-wishListButton-31"
        data-testid="wishlistButton"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          WISHLIST
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
